### ISSUE TITLE
Summer 2016 Scale Run

### ISSUE THREAD

**Issue by annawoodard** [matz-e/lobster#338]
Let's collect information here as we prepare for the scale run. I think it's a good idea to go ahead and start posting tests as we run them. It will be helpful to get as many eyes on them as possible.

| configuration | plots | logs | notes
| ------------ | ------------- | -------- | -----
bandwidth-limited run (ntuplizing configuration) | |
specify resources manually, WQ doesn't update (`WORK_QUEUE_ALLOCATION_MODE_FIXED` with resources specified) | |
WQ figures everything out (anything but `WORK_QUEUE_ALLOCATION_MODE_FIXED`, without resources specified) | [plots](http://www.crc.nd.edu/~awoodard/lobster/ttZ/v3/a/) |`/tmpscratch/users/awoodard/ttZ/v3/a/` | trying `max_throughput`, for now |
specify a starting point, WQ updates (anything but `WORK_QUEUE_ALLOCATION_MODE_FIXED`, with partial or all resources specified) | |
no categories (`WORK_QUEUE_ALLOCATION_MODE_FIXED`, without resources specified, and without `t.specify_category`) | |

**Comment by annawoodard** [comment]
For the tests that require != `WORK_QUEUE_ALLOCATION_MODE_FIXED`, we might not have enough time to test all of the allocation possibilities, so we need to decide which to test first between `WORK_QUEUE_ALLOCATION_MODE_MAX`, `WORK_QUEUE_ALLOCATION_MODE_MIN_WASTE`, and `WORK_QUEUE_ALLOCATION_MODE_MAX_THROUGHPUT`.

**Comment by annawoodard** [comment]
I've posted my first test. So far it isn't working (all `gs_step` jobs perpetually fail due to exhausted resources.) I'm working with @btovar to figure out what's going on.

**Comment by annawoodard** [comment]
Notes from today's meeting: `WORK_QUEUE_ALLOCATION_MODE_MAX`, `WORK_QUEUE_ALLOCATION_MODE_MIN_WASTE`, and `WORK_QUEUE_ALLOCATION_MODE_MAX_THROUGHPUT` will all be similar for these conditions. Let's plan to use `WORK_QUEUE_ALLOCATION_MODE_MAX_THROUGHPUT` for tests 3 and 4 above. We'll plan to do the bandwidth test first, followed by whatever we think will maximize resource utilization-- I'd say test 4. We'll finish up the other tests, possibly after people are allowed back on the queue, expecting that, as usual, we'll have some time before people really start coming back.

**Comment by annawoodard** [comment]
@matz-e, @klannon: We'll need to decide if we want to use `work_queue_factory --autosize` for the tests. I think that's definitely the easiest way to make sure we're getting every last available core. Is there any reason we'd need uniform-sized workers?

**Comment by klannon** [comment]
Yes, I think we should definitely try the `autosize` option.  I think it would be best to do that ASAP, maybe after we have the first run where we feel confident that Lobster and WQ are playing nice together?

**Comment by annawoodard** [comment]
Yes-- the last time I tried to use it, I didn't feel that confident that I understood what was going on. But I wanted to wait until we get everything else smoothed out before throwing another variable in the mix. The current test I'm running with the latest from @btovar is looking great so far, so I think we might be there, but I don't want to jinx it!

**Comment by klannon** [comment]
Does it make sense to build a list here or elsewhere of all the real work that we want to do using the upcoming computing so that if things go well we don't run out of useful samples?  Obviously, if we need to we can just recycle requests, but if things are going well, it would be nice to generate tens of millions of actually useful MC.

**Comment by annawoodard** [comment]
That's a good idea, I think we should put it here, to keep everything in the same place. As far as production goes, I was planning to just make as much as I could of ttZ, ttW, and ttH. I think the more I can make, the better. But of course we can add other MC configs in as well if you or @matz-e have things to run.

**Comment by klannon** [comment]
From my point of view, those are the important ones, although for ttH, I would do the non-bb sample as the top priority.  Ideally, we'd do this with CMSSW 8x (well for DR).  If someone doesn't beat me to it (and please be my guest!), I will try to dig out the appropriate requests from McM for the new campaign for those three to make sure we're making the must up-to-date stuff.

**Comment by annawoodard** [comment]
If you could dig those out, that would be great. I still have a few things to finish up before the April 30th deadline for freezing the code that I'd like to look at first.

**Comment by klannon** [comment]
OK, so this is a "progress" report on tracking down the relevant configs just so that my current progress isn't lost if I get overwhelmed.  What I'm trying to do is find the appropriate links in McM for the official configs for DR80 samples.  The idea being if we can validate our ability to run the most recent possible configuration, it will last as long as possible.

For reference, the general flow of state of the art MC production is something like:
 1. **LHE from RunIIWinter15wmLHE**
 2. **GS from RunIISummer15GS**
 3. **DR from RunIISpring16DR80**
 4. **RunIISpring16MiniAODv1** (There is one request in **RunIISpring16DRMiniAODv2**.  I don't know if we want to jump directly to v2 assuming that's where everything will go, or whether we want to stick with v1 because that's what people are running right now?)

Here's what we have for ttW and ttZ:

**ttZ:** [Chain](https://cms-pdmv.cern.ch/mcm/chained_requests?prepid=HIG-chain_RunIIWinter15wmLHE_flowLHE2Summer15GS_flowRunIISpring16DR80PU2016_flowRunIISpring16MiniAODv1-00027&page=0)
 1. LHE: [HIG-RunIIWinter15wmLHE-01035](https://cms-pdmv.cern.ch/mcm/requests?prepid=HIG-RunIIWinter15wmLHE-01035)
 2. GS:  [HIG-RunIISummer15GS-01168](https://cms-pdmv.cern.ch/mcm/requests?prepid=HIG-RunIISummer15GS-01168&page=0)
 3. DR80:  [HIG-RunIISpring16DR80-00045](https://cms-pdmv.cern.ch/mcm/requests?prepid=HIG-RunIISpring16DR80-00045&page=0)
 4. MiniAODV1:  [HIG-RunIISpring16MiniAODv1-00045](https://cms-pdmv.cern.ch/mcm/requests?prepid=HIG-RunIISpring16MiniAODv1-00045&page=0)

**ttW:** [Chain](https://cms-pdmv.cern.ch/mcm/chained_requests?prepid=HIG-chain_RunIIWinter15wmLHE_flowLHE2Summer15GS_flowRunIISpring16DR80PU2016_flowRunIISpring16MiniAODv1-00058&page=0)
 1. LHE: [HIG-RunIIWinter15wmLHE-01036](https://cms-pdmv.cern.ch/mcm/requests?prepid=HIG-RunIIWinter15wmLHE-01036&page=0)
 2. GS:  [HIG-RunIISummer15GS-01166](https://cms-pdmv.cern.ch/mcm/requests?prepid=HIG-RunIISummer15GS-01166&page=0)
 3. DR80:  [HIG-RunIISpring16DR80-00069](https://cms-pdmv.cern.ch/mcm/requests?prepid=HIG-RunIISpring16DR80-00069&page=0)
 4. MiniAODV1:  [HIG-RunIISpring16MiniAODv1-00069](https://cms-pdmv.cern.ch/mcm/requests?prepid=HIG-RunIISpring16MiniAODv1-00069&page=0)

**ttH(non-bb):** (This doesn't seem to have been put into 80x yet.)
 1. LHE: [HIG-RunIIWinter15wmLHE-00195](https://cms-pdmv.cern.ch/mcm/requests?prepid=HIG-RunIIWinter15wmLHE-00195&page=0)
 2. GS: [HIG-RunIISummer15GS-00176](https://cms-pdmv.cern.ch/mcm/requests?prepid=HIG-RunIISummer15GS-00176&page=0)

Then we could steal the DR and MiniAOD from above?

How are you guys converting the McM requests into Lobster?  Is there something automated?  When I last did it, I did it manually.  Is there something fancier now?

**Comment by matz-e** [comment]
Finally got back and settled in… I've converted it manually so far. I'd say an hour or two of work is all it takes. I looked into the python API for MCM, but it's pretty hopeless, I think, since there does not seem to be a very good way to download everything and detect chained steps?

Plus, the API is best used from CERN, it'd be a pain to get it working at ND.

**Comment by matz-e** [comment]
Added my workflow that I ran over the last couple of days…

**Comment by annawoodard** [comment]
@matz-e: Do we understand the large failures in your run?

**Comment by matz-e** [comment]
Dataset vs. pset is the main issue, I think. I added a whole bunch of datasets, and I need to follow up what exactly is going wrong with these and then fix the config.

**Comment by annawoodard** [comment]
Let's start listing anything here about the plots that needs to be understood before the scale run.

1. Why are there large failures with some datasets in the ntuplizing config?
2. Why are all the plots under 'Task Resources Requested' zero?

**Comment by matz-e** [comment]
Well, my quick shots, I'll investigate more next week:

1. These datasets haven't been processed for a while, they are not in my canoncial TTL set, but the new TLL… as such I'll see what's going on. I could always take them out.
2. I removed all resource requests from my configuration, I think the plots show what WQ received as a request from the user. Which is not what I intended to plot… I wanted to have the resource allocation evolution over time as WQ assigns resources to the tasks… not sure how to obtain that but from the transaction log? @btovar, can we get the current resource allocation that WQ thinks is optimal via the API?

**Comment by annawoodard** [comment]
@matz-e: Can you post your `chirp_server` command? I want to understand why I have higher stageout times.

**Comment by annawoodard** [comment]
Another mystery to solve: why is the event count lagging behind the units processed count? It does not seem to be filtering as the events read is also lagging:
![screen shot 2016-05-06 at 4 56 40 pm](https://cloud.githubusercontent.com/assets/5114833/15085972/a8572e8a-13ab-11e6-94da-4931d727f6b1.png)

**Comment by matz-e** [comment]
Actually, I did not use a chirp server at all, since I had weird issues with my stageout directory not being writeable (locally logging into the chirp server and copying files over worked, though).

**Comment by annawoodard** [comment]
Hmmm. We should have a look at that before the scale run. What did you use for stageout?

**Comment by matz-e** [comment]
SRM, but I had to fix the xrootd URL that Lobster autodetected.

**Comment by annawoodard** [comment]
(Sorry, I recall now that I can just check the config linked to from your plots. That's nice!)

**Comment by matz-e** [comment]
That's what I used to fix the URL:
`./lobster.py configure "storage.output[2]='root://deepthought.crc.nd.edu//store/user/matze/ttH/v27/'" /tmpscratch/users/matze/tau_v27/`

Without it, it would point to NDCMS, which is just weird, because that shouldn't be resolved?

**Comment by annawoodard** [comment]
Hmm. Maybe @khurtado knows?

We can just hardcode the xrootd url in the config, no?

**Comment by matz-e** [comment]
Yes, that's a possibility. I should actually do that…

**Comment by khurtado** [comment]
@matz-e: It should indeed be resolved. The SE points to deepthought in SiteDB. What were you doing before? `root://T3_US_Notredame//store/user/matze/ttH/v27`? Could you point me to the auto-detect function in Lobster that resolves to NDCMS?

**Comment by matz-e** [comment]
@khurtado I think my memory must be mistaken… Just tried this again, and I got the `xrootd.unl.edu` redirector, which I think matches the configuration, but does not resolve and still has to be fixed manually.

**Comment by matz-e** [comment]
For reference, [here](https://github.com/matz-e/lobster/blob/master/lobster/se.py#L416) we try to get the right server name from SITECONF.

**Comment by matz-e** [comment]
I guess we should switch to SiteDB?

**Comment by klannon** [comment]
(FYI: discussion of SiteDB moved to #354.  In any case, we've decided, I think, to postpone that until after the scale test and just manually specify `root:deepthought....` instead of `root:T3_US_NotreDame...` for the scale test.)

Where do we stand for the scale test?  Can we run tests 1 and 3 successfully?  Can we do it with the XRootD cache in the chain?

**Comment by matz-e** [comment]
I can kick off another test for (1), since it looks like a lot of stuff got skipped and I did not max out the campus bandwidth during my test… there should be plenty data to pull in.

As far as the cache, do we have some URLs already that we could try splicing into the processing/name resolution chain? I'm inclined to first hack the SITECONF we ship and then implement a setting if that's the right way to go.

**Comment by annawoodard** [comment]
Can @khurtado give us an update on the status of the cache?

**Comment by klannon** [comment]
Oops!  You guys were left of the e-mail chain.  Message sent...

**Comment by annawoodard** [comment]
I am finishing up getting the new configs prepared by @klannon running with my hacked LHE code. That should be finished this morning. Then my plan was to run a short test each for test types 2-5 (1,000 or 10,000 events). Then hopefully be ready for a test of type 3 or 4 where I try to max everything out by tomorrow. I'm tempted to run the small tests in parallel, because one doesn't fill up the queue, but I'm wary of introducing any artificial problems. I've just updated with a short type 4 test (old config). The tail is pretty long because of merging. If either/both of you could have a look at the plots and let me know if you see anything that doesn't look right before I submit the next batch of tests that would be helpful.

**Comment by matz-e** [comment]
Decreasing the merge size for these tests would probably help? Just peeked into one of your configs (v46), and it seems like you have 3.5 GB as merge size in there?

**Comment by annawoodard** [comment]
@matz-e: Good point. I deliberately left it high because I was curious how long it would take, but I should turn it down for the next short tests.

**Comment by klannon** [comment]
On which steps do you apply merging?  Presumably you only need it for the output of RECO (e.g. AOD) and miniAOD, right?  For miniAOD, I think 3.5 GB is actually pretty large.  I think 1 GB is plenty merged, if that helps.

**Comment by matz-e** [comment]
I have been merging all steps, basically to test that inserting merged files works properly. Now I guess I could just delete some intermediate outputs…

**Comment by annawoodard** [comment]
I'm only applying merging for RECO and miniAOD step. Good point, that I can reduce the merge size to 1 GB. Thanks!

Does a lumi size of 100 events seem ok to you guys? That's what I normally see in central production. It results in a mean task runtime of ~90 minutes for gen-sim (one lumi per task). I figured being bigger would reduce the per-lumi storage space, but now that I think about it, I guess that's probably tiny.

**Comment by matz-e** [comment]
I think that's a good size. The optimal runtime in our paper is around that point (eviction-wise).

**Comment by matz-e** [comment]
For your v45: I have the hunch that while merging the number of events read/written gets set to 0, since it comes from successful / merged tasks (could add a merging status in there). If you have a merge task in task in the queue that somehow is not getting processed, that'd explain why that run is stuck…

**Comment by annawoodard** [comment]
Thanks!

**Comment by annawoodard** [comment]
Unless anyone has an objection, for my next short tests I'm going to disable merging. I think that we've established that it works, and I'll be able to get through the whole chain faster without it. I'll re-instate it for the bigger tests tomorrow.

**Comment by annawoodard** [comment]
Short update from me:

I think I have the new spring16DR80 configurations working properly with my six-dimensional madgraph LHE scan, and I've gone through each test type to make sure that we're actually able to run it in Lobster. I've posted small runs for test types 2-5. (Types 2 and 5 are still running.) After they're all finished, any help I can get looking over the plots tomorrow to try and spot anything that looks fishy will be really helpful (@klannon, @matz-e, @khurtado if you have a few minutes after the XrootD caching is squared away.) @btovar, if you have a few minutes tomorrow, it would be good if you could quickly double check the logs and make sure you think we're writing out everything you need, and also that test type 5 looks implemented correctly to you.

I'm about to start a larger scale test type 4 with merging re-enabled and pointing at @khurtado's test XrootD cache to run overnight.

**Comment by btovar** [comment]
@annawoodard: which logs should I be double-checking?

**Comment by matz-e** [comment]
Damn. I updated the table earlier with my 2nd ntuplization. I hope I didn't overwrite anything there…

**Comment by btovar** [comment]
Are you also plotting stats.total_exhausted_retries?

**Comment by annawoodard** [comment]
@btovar: Thanks for the feedback! We are plotting task.exhausted_retries, that should be the same, no? [example](http://www.crc.nd.edu/~awoodard/lobster/ttW/v6/a/gs/good-exhausted-attempts-prof.png)

I have updated the plot links which were overwritten. If you could check test 5 to see that you think it's implemented correctly, that'd be really helpful!

**Comment by btovar** [comment]
The average, I assume?

**Comment by annawoodard** [comment]
It's the total from all tasks for that time bin.

**Comment by btovar** [comment]
So in your example it means zero exhaustion attempts?

**Comment by annawoodard** [comment]
Yes.

**Comment by klannon** [comment]
Proposal for this weekend:

I still need to confirm with the CRC, but let's assume that we get a shot with the full system sometime on Saturday.  We have a number of things we want to test, but after discussion, I think this seems like the most sensible plan:

 1.  **Start with the big MC scale test run by @annawoodard.**  If the xrootd cache is working, include it.  If it's not working, just use the regular network link (aka 10 Gb/s).  **Goals:** (a) Get another set of plots running at scale, confirm that we haven't regressed with all of our development since Jan.  (b) Determine whether we see a significantly higher throughput now that we're (hopefully) making use of multicore processing.  (c) Make a list of what breaks when we run at this scale with the current configuration.  How close are we to optimal throughput?  What are the bottlenecks that prevent us from utilizing the full scale of resources.  **Requirements:** Can only happen during the CRC down time when the system almost completely empty.
 2.  **Network/cache stress test by @matz-e running ntuple jobs.** **Goals:** (a) Determine whether we can push the network to 20 Gb/s with two caching proxies.  (b) Understand the scaling.  Is it linear? (e.g. if I saturate a cluster with two proxy caches by using 8000 cores, does that mean I saturate just one with 4000?  **Requirements:** Needs the proxy cache to be working (isn't currently).  Might not be able to happen right away during the outage.  Can happen after the downtime if the system is lightly enough utilized, but if we can do it before the end of the downtime, that would be ideal.
 3.  **Comparing Work Queue Allocation strategies.**  The basic idea is to try the four different modes of Work Queue resource allocations and collect data on them for @btovar's paper.  Presumably we want to make each one of these runs as similar as possible in terms of what gets processed, which means we probably want to start the Lobster project over again each time.  Probably the safest thing is to pick one dataset and break it into four pieces and do one piece with each mode.  **Goals:** Accumulate enough data with each strategy that @btovar is happy.  **Requirements:** I think this requires neither the full system (so doesn't have to be done during the outage) nor the proxy cache if it's not working, so we should just do it as soon as possible after the outage ends.  Technically, we'll have data from one mode already during the big scale test, so @btovar should look at that and decide if that's a good enough starting point.

Possible timeline:
**Sat through Sun.**: Test 1 (big scale run)
**Mon. morning Europe:** Test 2
** Rest of next week:** Test 3

Please give feedback and we'll revise until the plan is satisfactory for everyone.
